File: polymarket/models/calibration.py
"""
Probability calibration and model evaluation
Implements Brier score and calibration analysis
"""

import numpy as np
import pandas as pd
from typing import Tuple


class CalibrationAnalyzer:
    """Analyzes and improves probability calibration"""

    @staticmethod
    def brier_score(predicted_probs: np.ndarray, outcomes: np.ndarray) -> float:
        """
        Calculate Brier score - mean squared error between predictions and outcomes

        Args:
            predicted_probs: Array of probability predictions [0, 1]
            outcomes: Array of binary outcomes [0, 1]

        Returns:
            Brier score (lower is better, 0 is perfect)
        """
        return np.mean((predicted_probs - outcomes) ** 2)

    @staticmethod
    def reliability_diagram(
        predicted_probs: np.ndarray,
        outcomes: np.ndarray,
        n_bins: int = 10
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        Create reliability diagram data for calibration visualization

        Args:
            predicted_probs: Array of probability predictions
            outcomes: Array of binary outcomes
            n_bins: Number of bins for binning predictions

        Returns:
            Tuple of (bin_edges, bin_means, observed_frequencies)
        """
        bin_edges = np.linspace(0, 1, n_bins + 1)
        bin_means = []
        observed_frequencies = []

        for i in range(n_bins):
            mask = (predicted_probs >= bin_edges[i]) & (predicted_probs < bin_edges[i + 1])
            if mask.sum() > 0:
                bin_means.append(predicted_probs[mask].mean())
                observed_frequencies.append(outcomes[mask].mean())
            else:
                bin_means.append((bin_edges[i] + bin_edges[i + 1]) / 2)
                observed_frequencies.append(np.nan)

        return bin_edges, np.array(bin_means), np.array(observed_frequencies)

    @staticmethod
    def calibration_metrics(
        predicted_probs: np.ndarray,
        outcomes: np.ndarray
    ) -> dict:
        """
        Calculate comprehensive calibration metrics

        Args:
            predicted_probs: Array of probability predictions
            outcomes: Array of binary outcomes

        Returns:
            Dictionary with calibration metrics
        """
        brier = CalibrationAnalyzer.brier_score(predicted_probs, outcomes)

        # Expected calibration error
        ece = CalibrationAnalyzer.expected_calibration_error(predicted_probs, outcomes)

        # Sharpness
        sharpness = np.std(predicted_probs)

        # Resolution
        base_rate = np.mean(outcomes)
        resolution = np.mean((predicted_probs - base_rate) ** 2)

        # Coverage
        in_credible = (predicted_probs >= 0) & (predicted_probs <= 1)
        coverage = np.mean(in_credible)

        return {
            "brier_score": brier,
            "expected_calibration_error": ece,
            "sharpness": sharpness,
            "resolution": resolution,
            "coverage": coverage,
        }

    @staticmethod
    def expected_calibration_error(
        predicted_probs: np.ndarray,
        outcomes: np.ndarray,
        n_bins: int = 10
    ) -> float:
        """
        Calculate expected calibration error

        Args:
            predicted_probs: Array of probability predictions
            outcomes: Array of binary outcomes
            n_bins: Number of bins

        Returns:
            Expected calibration error
        """
        bin_edges, bin_means, observed_freq = CalibrationAnalyzer.reliability_diagram(
            predicted_probs, outcomes, n_bins
        )

        # Calculate ECE as weighted average of bin calibration errors
        bin_sizes = np.histogram(predicted_probs, bin_edges)[0]
        bin_sizes = bin_sizes[bin_sizes > 0] / len(predicted_probs)

        valid_bins = ~np.isnan(observed_freq[bin_sizes > 0])
        bin_means_valid = bin_means[bin_sizes > 0][valid_bins]
        observed_freq_valid = observed_freq[bin_sizes > 0][valid_bins]
        bin_sizes_valid = bin_sizes[valid_bins]

        ece = np.sum(bin_sizes_valid * np.abs(bin_means_valid - observed_freq_valid))

        return ece

    @staticmethod
    def temperature_calibration(
        predicted_temps: np.ndarray,
        observed_temps: np.ndarray
    ) -> dict:
        """
        Calibration analysis for continuous temperature predictions

        Args:
            predicted_temps: Array of predicted temperatures
            observed_temps: Array of observed temperatures

        Returns:
            Dictionary with calibration metrics
        """
        mae = np.mean(np.abs(predicted_temps - observed_temps))
        rmse = np.sqrt(np.mean((predicted_temps - observed_temps) ** 2))
        r_squared = 1 - (
            np.sum((observed_temps - predicted_temps) ** 2) /
            np.sum((observed_temps - np.mean(observed_temps)) ** 2)
        )

        # Prediction interval coverage
        std_error = np.std(predicted_temps - observed_temps)
        coverage_68 = np.mean(
            np.abs(predicted_temps - observed_temps) <= std_error
        )
        coverage_95 = np.mean(
            np.abs(predicted_temps - observed_temps) <= 1.96 * std_error
        )

        return {
            "mae": mae,
            "rmse": rmse,
            "r_squared": r_squared,
            "coverage_68pct": coverage_68,
            "coverage_95pct": coverage_95,
        }

File: src/models/calibration.rs
use crate::utils::{mean, std_dev};

pub struct CalibrationAnalyzer;

impl CalibrationAnalyzer {
    pub fn brier_score(predicted_probs: &[f64], outcomes: &[f64]) -> f64 {
        if predicted_probs.is_empty() || predicted_probs.len() != outcomes.len() {
            return 0.0;
        }

        predicted_probs
            .iter()
            .zip(outcomes.iter())
            .map(|(p, y)| {
                let d = p - y;
                d * d
            })
            .sum::<f64>()
            / predicted_probs.len() as f64
    }

    pub fn reliability_diagram(
        predicted_probs: &[f64],
        outcomes: &[f64],
        n_bins: usize,
    ) -> (Vec<f64>, Vec<f64>, Vec<f64>) {
        let edges: Vec<f64> = (0..=n_bins).map(|i| i as f64 / n_bins as f64).collect();
        let mut bin_means = Vec::with_capacity(n_bins);
        let mut observed = Vec::with_capacity(n_bins);

        for i in 0..n_bins {
            let lo = edges[i];
            let hi = edges[i + 1];
            let mut probs = Vec::new();
            let mut outs = Vec::new();

            for (p, y) in predicted_probs.iter().zip(outcomes.iter()) {
                if *p >= lo && *p < hi {
                    probs.push(*p);
                    outs.push(*y);
                }
            }

            if probs.is_empty() {
                bin_means.push((lo + hi) / 2.0);
                observed.push(f64::NAN);
            } else {
                bin_means.push(mean(&probs));
                observed.push(mean(&outs));
            }
        }

        (edges, bin_means, observed)
    }

    pub fn expected_calibration_error(
        predicted_probs: &[f64],
        outcomes: &[f64],
        n_bins: usize,
    ) -> f64 {
        if predicted_probs.is_empty() || predicted_probs.len() != outcomes.len() {
            return 0.0;
        }

        let (edges, bin_means, obs) = Self::reliability_diagram(predicted_probs, outcomes, n_bins);
        let total = predicted_probs.len() as f64;
        let mut ece = 0.0;

        for i in 0..n_bins {
            let lo = edges[i];
            let hi = edges[i + 1];
            let count = predicted_probs
                .iter()
                .filter(|p| **p >= lo && **p < hi)
                .count() as f64;

            if count > 0.0 && obs[i].is_finite() {
                ece += (count / total) * (bin_means[i] - obs[i]).abs();
            }
        }

        ece
    }

    pub fn calibration_metrics(predicted_probs: &[f64], outcomes: &[f64]) -> std::collections::HashMap<String, f64> {
        let mut out = std::collections::HashMap::new();
        let brier = Self::brier_score(predicted_probs, outcomes);
        let ece = Self::expected_calibration_error(predicted_probs, outcomes, 10);

        let sharpness = std_dev(predicted_probs, 0);
        let base_rate = mean(outcomes);
        let resolution = predicted_probs
            .iter()
            .map(|p| {
                let d = p - base_rate;
                d * d
            })
            .sum::<f64>()
            / predicted_probs.len().max(1) as f64;

        let coverage = if predicted_probs.is_empty() {
            0.0
        } else {
            predicted_probs
                .iter()
                .filter(|p| **p >= 0.0 && **p <= 1.0)
                .count() as f64
                / predicted_probs.len() as f64
        };

        out.insert("brier_score".to_string(), brier);
        out.insert("expected_calibration_error".to_string(), ece);
        out.insert("sharpness".to_string(), sharpness);
        out.insert("resolution".to_string(), resolution);
        out.insert("coverage".to_string(), coverage);
        out
    }

    pub fn temperature_calibration(
        predicted_temps: &[f64],
        observed_temps: &[f64],
    ) -> std::collections::HashMap<String, f64> {
        let n = predicted_temps.len().min(observed_temps.len());
        if n == 0 {
            return std::collections::HashMap::new();
        }

        let mut abs_errors = Vec::with_capacity(n);
        let mut sq_errors = Vec::with_capacity(n);
        let mut residuals = Vec::with_capacity(n);

        for i in 0..n {
            let err = predicted_temps[i] - observed_temps[i];
            abs_errors.push(err.abs());
            sq_errors.push(err * err);
            residuals.push(err);
        }

        let mae = mean(&abs_errors);
        let rmse = mean(&sq_errors).sqrt();

        let observed_mean = mean(&observed_temps[..n]);
        let ss_res = sq_errors.iter().sum::<f64>();
        let ss_tot = observed_temps[..n]
            .iter()
            .map(|y| {
                let d = y - observed_mean;
                d * d
            })
            .sum::<f64>();
        let r_squared = if ss_tot > 0.0 { 1.0 - ss_res / ss_tot } else { 0.0 };

        let std_error = std_dev(&residuals, 0);
        let coverage_68 = residuals
            .iter()
            .filter(|e| e.abs() <= std_error)
            .count() as f64
            / n as f64;
        let coverage_95 = residuals
            .iter()
            .filter(|e| e.abs() <= 1.96 * std_error)
            .count() as f64
            / n as f64;

        let mut out = std::collections::HashMap::new();
        out.insert("mae".to_string(), mae);
        out.insert("rmse".to_string(), rmse);
        out.insert("r_squared".to_string(), r_squared);
        out.insert("coverage_68pct".to_string(), coverage_68);
        out.insert("coverage_95pct".to_string(), coverage_95);
        out
    }
}
